tempesta_fw/t/unit/tfw_fuzzer.c

```diff
@@ -0,0 +1,563 @@
+		s1[len + i] = a_body[((i + gen_vector[CONTENT_LENGTH].i +
+				       gen_vector[BODY_SIZE].i + len) ^
+				       (a_body[i] + len)) % sizeof(a_body)];
+
+	err = kstrtoul(content_length[gen_vector[CONTENT_LENGTH].i].s,
+			10, &cont_len);
+	if (err)
+		return FUZZ_INVALID;
+	if (cont_len > body_size)
+		return FUZZ_INVALID;
+	if (type == FUZZ_REQ)
+		return FUZZ_VALID;
+
+	rc = resp_code[gen_vector[RESP_CODE].i].s;
+	if (!strstr(rc, "100") || !strstr(rc, "101") || !strstr(rc, "102") ||
+			!strstr(rc, "204") || !strstr(rc, "304"))
+		return FUZZ_INVALID;
+
+	return FUZZ_VALID;
+}
+
+static int add_duplicates(char *s1, int n, int(*add_func)(char *s1))
+{
+	int i, tmp, v = 0;
+
+	for (i = 0; i < gen_vector[DUPLICATES].i; ++i) {
+		tmp = gen_vector[n].i;
+
+		gen_vector[n].i = (i + gen_vector[n].i +
+				   gen_vector[SPACES].i) %
+				  gen_vector[n].max;
+		v |= (*add_func)(s1);
+		strcat(s1, "\r\n");
+		gen_vector[n].i = tmp;
+	}
+
+	if (gen_vector[n].singular && i > 0)
+		return FUZZ_INVALID;
+	return v;
+}
+
+/* Returns:
+ * FUZZ_VALID if the result is a valid request,
+ * FUZZ_INVALID if it's invalid,
+ * FUZZ_END if the request sequence is over.
+ * `move` is how many gen_vector's elements should be changed
+ * each time a new request is generated, should be >= 1. */
+int fuzz_gen(char *str, int move, int type)
+{
+	int i, ret, v = 0;
+
+	if (str == NULL)
+		return -EINVAL;
+	if (move < 1)
+		return -EINVAL;
+	*str = '\0';
+
+	if (type == FUZZ_REQ) {
+		v |= add_method(str);
+		add_spaces(str);
+
+		add_uri_path_start(str);
+		if (!gen_vector[URI_PATH_DEPTH].i)
+			addch(str, '/');
+		for (i = 0; i < gen_vector[URI_PATH_DEPTH].i; ++i) {
+			addch(str, '/');
+			v |= add_uri_file(str);
+		}
+		add_spaces(str);
+	}
+
+	strcat(str, "HTTP/");
+	v |= add_http_ver(str);
+	if (type == FUZZ_RESP) {
+		add_spaces(str);
+		v |= add_resp_code(str);
+	}
+	strcat(str, "\r\n");
+
+	v |= add_connection(str);
+	strcat(str, "\r\n");
+	v |= add_duplicates(str, CONNECTION, &add_connection);
+
+	v |= add_user_agent(str);
+	strcat(str, "\r\n");
+	v |= add_duplicates(str, USER_AGENT, &add_user_agent);
+
+	v |= add_host(str);
+	strcat(str, "\r\n");
+	v |= add_duplicates(str, HOST, &add_host);
+
+	v |= add_x_forwarded_for(str);
+	strcat(str, "\r\n");
+	v |= add_duplicates(str, X_FORWARDED_FOR, &add_x_forwarded_for);
+
+	v |= add_content_type(str);
+	strcat(str, "\r\n");
+	v |= add_duplicates(str, CONTENT_TYPE, &add_content_type);
+
+	v |= add_content_length(str);
+	strcat(str, "\r\n");
+	v |= add_duplicates(str, CONTENT_LENGTH, &add_content_length);
+
+	v |= add_transfer_encoding(str);
+	strcat(str, "\r\n");
+	v |= add_duplicates(str, TRANSFER_ENCODING, &add_transfer_encoding);
+
+	strcat(str, "\r\n");
+
+	v |= add_body(str, type);
+
+	for (i = 0; i < move; i++) {
+		ret = gen_vector_move((N_FIELDS - 1 + i * ((i % 2) ? 3 : 2)) % N_FIELDS);
+		if (ret == FUZZ_END)
+			break;
+	}
+	if (v & FUZZ_INVALID)
+		return FUZZ_INVALID;
+	return ret;
```

Comment:
#150 ends by the list of papers with many HTTP attacks and no one of them is implemented in the fuzzer. I think most of them can be encoded as special values for header values (e.g. request smuggling should be just a one-two special values for any of the headers). Please review the papers for examples of HTTP attacks and accurately implement them in the fuzzer.